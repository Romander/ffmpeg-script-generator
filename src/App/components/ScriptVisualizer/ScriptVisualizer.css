.script-visualizer-item {
    white-space: pre;
}
