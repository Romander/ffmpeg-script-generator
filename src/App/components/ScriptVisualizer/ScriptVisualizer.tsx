import React from "react";
import { IFFmpegInputParam } from "../../types";
import { ScriptVisualizerView } from "./components/ScriptVisualizerView/ScriptVisualizerView";

interface IScriptVisualizerProps {
  enableNvideaDecode: boolean;
  enableNvideaEncode: boolean;
  inputs: Array<IFFmpegInputParam>;
  filterComplexVideosSettings?: Array<string>;
  filterComplexAudiosSettings?: Array<string>;
  xfades?: Array<string>;
  acrossfades?: Array<string>;
  mapping: string;
  outputFilename: string;
}

function ScriptVisualizer(props: IScriptVisualizerProps) {
  return (
    <ScriptVisualizerView>
      <div>ffmpeg \</div>
      <div>-vsync 0 \</div>
      {props.enableNvideaDecode && <div>-c:v h264_cuvid \</div>}
      {props.inputs.map((input) => (
        <div>-i "{input.name}" \</div>
      ))}
      <div>
        -filter_complex \
        {props.filterComplexVideosSettings?.map((setting, index) => (
          <div>
            {index === 0 ? '"' : null}
            {setting} \
          </div>
        ))}
        {props.filterComplexAudiosSettings?.map((setting) => (
          <div>{setting} \</div>
        ))}
      </div>
      <div>
        {props.xfades?.map((xfade, index) => (
          <div>
            {xfade}
            {props.xfades?.length && index === props.xfades?.length - 1
              ? ";"
              : null}
            \
          </div>
        ))}
        {props.acrossfades?.map((acrossfade, index) => (
          <div>
            {acrossfade}
            {props.acrossfades?.length &&
            index === props.acrossfades?.length - 1
              ? '"'
              : null}
            \
          </div>
        ))}
      </div>
      {props.enableNvideaEncode && <div>-c:v h264_nvenc \</div>}
      <div>-b:v 10M \</div>
      <div>
        {props.mapping} "{props.outputFilename}" \
      </div>
    </ScriptVisualizerView>
  );
}

export { ScriptVisualizer };
