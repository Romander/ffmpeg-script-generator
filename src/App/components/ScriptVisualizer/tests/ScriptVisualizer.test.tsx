import * as uuid from 'uuid';
import renderer from 'react-test-renderer';
import { ScriptVisualizer } from '../ScriptVisualizer';

const getProps = (newProps?: any) => ({
    enableNvideaDecode: false,
    enableNvideaEncode: false,
    inputs: [
        { id: uuid.v4(), name: 'F1.AVI', duration: 10 },
        { id: uuid.v4(), name: 'F2.AVI', duration: 10 },
    ],
    filterComplexVideosSettings: ['[0]settb=AVTB[0:v];', '[1]settb=AVTB[1:v];'],
    filterComplexAudiosSettings: ['[0]atrim=0:10[0:a];', '[1]atrim=0:10[1:a];'],
    xfades: [
        '[0:v][1:v]xfade=transition=fade:duration=1:offset=9,format=yuv420p[video]',
    ],
    acrossfades: ['[0:a][1:a]acrossfade=d=1:c1=tri:c2=tri[audio]'],
    mapping: '-map "[audio]" -map "[video]"',
    outputFilename: 'out.mp4',
    ...newProps,
});

test('ScriptVisualizer should render correctly', () => {
    const component = renderer.create(<ScriptVisualizer {...getProps()} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('ScriptVisualizer should render with enableNvideaDecode setting', () => {
    const component = renderer.create(
        <ScriptVisualizer {...getProps({ enableNvideaDecode: true })} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('ScriptVisualizer should render with enableNvideaEncode setting', () => {
    const component = renderer.create(
        <ScriptVisualizer {...getProps({ enableNvideaEncode: true })} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('ScriptVisualizer should render with enableNvideaDecode and enableNvideaEncode setting', () => {
    const component = renderer.create(
        <ScriptVisualizer
            {...getProps({
                enableNvideaDecode: true,
                enableNvideaEncode: true,
            })}
        />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('ScriptVisualizer should render without acrossfades', () => {
    const component = renderer.create(
        <ScriptVisualizer {...getProps({ acrossfades: null })} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
