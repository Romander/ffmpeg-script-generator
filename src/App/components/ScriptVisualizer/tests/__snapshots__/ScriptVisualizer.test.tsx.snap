// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScriptVisualizer should render correctly 1`] = `
<div
  className="script-visualizer-view"
>
  <div>
    ffmpeg \\
  </div>
  <div>
    -vsync 0 \\
  </div>
  <div>
    -i "
    F1.AVI
    " \\
  </div>
  <div>
    -i "
    F2.AVI
    " \\
  </div>
  <div>
    -filter_complex \\
    <div>
      "
      [0]settb=AVTB[0:v];
       \\
    </div>
    <div>
      [1]settb=AVTB[1:v];
       \\
    </div>
    <div>
      [0]atrim=0:10[0:a];
       \\
    </div>
    <div>
      [1]atrim=0:10[1:a];
       \\
    </div>
  </div>
  <div>
    <div>
      [0:v][1:v]xfade=transition=fade:duration=1:offset=9,format=yuv420p[video]
      ;
      \\
    </div>
    <div>
      [0:a][1:a]acrossfade=d=1:c1=tri:c2=tri[audio]
      "
      \\
    </div>
  </div>
  <div>
    -b:v 10M \\
  </div>
  <div>
    -map "[audio]" -map "[video]"
     "
    out.mp4
    " \\
  </div>
</div>
`;

exports[`ScriptVisualizer should render with enableNvideaDecode and enableNvideaEncode setting 1`] = `
<div
  className="script-visualizer-view"
>
  <div>
    ffmpeg \\
  </div>
  <div>
    -vsync 0 \\
  </div>
  <div>
    -c:v h264_cuvid \\
  </div>
  <div>
    -i "
    F1.AVI
    " \\
  </div>
  <div>
    -i "
    F2.AVI
    " \\
  </div>
  <div>
    -filter_complex \\
    <div>
      "
      [0]settb=AVTB[0:v];
       \\
    </div>
    <div>
      [1]settb=AVTB[1:v];
       \\
    </div>
    <div>
      [0]atrim=0:10[0:a];
       \\
    </div>
    <div>
      [1]atrim=0:10[1:a];
       \\
    </div>
  </div>
  <div>
    <div>
      [0:v][1:v]xfade=transition=fade:duration=1:offset=9,format=yuv420p[video]
      ;
      \\
    </div>
    <div>
      [0:a][1:a]acrossfade=d=1:c1=tri:c2=tri[audio]
      "
      \\
    </div>
  </div>
  <div>
    -c:v h264_nvenc \\
  </div>
  <div>
    -b:v 10M \\
  </div>
  <div>
    -map "[audio]" -map "[video]"
     "
    out.mp4
    " \\
  </div>
</div>
`;

exports[`ScriptVisualizer should render with enableNvideaDecode setting 1`] = `
<div
  className="script-visualizer-view"
>
  <div>
    ffmpeg \\
  </div>
  <div>
    -vsync 0 \\
  </div>
  <div>
    -c:v h264_cuvid \\
  </div>
  <div>
    -i "
    F1.AVI
    " \\
  </div>
  <div>
    -i "
    F2.AVI
    " \\
  </div>
  <div>
    -filter_complex \\
    <div>
      "
      [0]settb=AVTB[0:v];
       \\
    </div>
    <div>
      [1]settb=AVTB[1:v];
       \\
    </div>
    <div>
      [0]atrim=0:10[0:a];
       \\
    </div>
    <div>
      [1]atrim=0:10[1:a];
       \\
    </div>
  </div>
  <div>
    <div>
      [0:v][1:v]xfade=transition=fade:duration=1:offset=9,format=yuv420p[video]
      ;
      \\
    </div>
    <div>
      [0:a][1:a]acrossfade=d=1:c1=tri:c2=tri[audio]
      "
      \\
    </div>
  </div>
  <div>
    -b:v 10M \\
  </div>
  <div>
    -map "[audio]" -map "[video]"
     "
    out.mp4
    " \\
  </div>
</div>
`;

exports[`ScriptVisualizer should render with enableNvideaEncode setting 1`] = `
<div
  className="script-visualizer-view"
>
  <div>
    ffmpeg \\
  </div>
  <div>
    -vsync 0 \\
  </div>
  <div>
    -i "
    F1.AVI
    " \\
  </div>
  <div>
    -i "
    F2.AVI
    " \\
  </div>
  <div>
    -filter_complex \\
    <div>
      "
      [0]settb=AVTB[0:v];
       \\
    </div>
    <div>
      [1]settb=AVTB[1:v];
       \\
    </div>
    <div>
      [0]atrim=0:10[0:a];
       \\
    </div>
    <div>
      [1]atrim=0:10[1:a];
       \\
    </div>
  </div>
  <div>
    <div>
      [0:v][1:v]xfade=transition=fade:duration=1:offset=9,format=yuv420p[video]
      ;
      \\
    </div>
    <div>
      [0:a][1:a]acrossfade=d=1:c1=tri:c2=tri[audio]
      "
      \\
    </div>
  </div>
  <div>
    -c:v h264_nvenc \\
  </div>
  <div>
    -b:v 10M \\
  </div>
  <div>
    -map "[audio]" -map "[video]"
     "
    out.mp4
    " \\
  </div>
</div>
`;

exports[`ScriptVisualizer should render without acrossfades 1`] = `
<div
  className="script-visualizer-view"
>
  <div>
    ffmpeg \\
  </div>
  <div>
    -vsync 0 \\
  </div>
  <div>
    -i "
    F1.AVI
    " \\
  </div>
  <div>
    -i "
    F2.AVI
    " \\
  </div>
  <div>
    -filter_complex \\
    <div>
      "
      [0]settb=AVTB[0:v];
       \\
    </div>
    <div>
      [1]settb=AVTB[1:v];
       \\
    </div>
    <div>
      [0]atrim=0:10[0:a];
       \\
    </div>
    <div>
      [1]atrim=0:10[1:a];
       \\
    </div>
  </div>
  <div>
    <div>
      [0:v][1:v]xfade=transition=fade:duration=1:offset=9,format=yuv420p[video]
      "
      \\
    </div>
  </div>
  <div>
    -b:v 10M \\
  </div>
  <div>
    -map "[audio]" -map "[video]"
     "
    out.mp4
    " \\
  </div>
</div>
`;
